clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -1196,16 +1196,31 @@ private long maybeDrainPendingCalls(long now) {
             long pollTimeout = Long.MAX_VALUE;
             log.trace("Trying to choose nodes for {} at {}", pendingCalls, now);
 
-            Iterator<Call> pendingIter = pendingCalls.iterator();
-            while (pendingIter.hasNext()) {
-                Call call = pendingIter.next();
+            List<Call> toRemove = new ArrayList<>();
+            // Using for-loop instead of iterator to avoid ConcurrentModificationException with following sequence:
+            // 1. In maybeDrainPendingCalls, pendingCalls.iterator() create a iterator.
+            // 2. In maybeDrainPendingCall, call.nodeProvider.provide() throws UnsupportedVersionException error.
+            // 3. In call.fail, runnable.pendingCalls.add modify the pendingCalls list.
+            // Using pendingCalls.size() to get the list size before the for-loop to avoid infinite loop.
+            // If call.fail keeps adding the call to pendingCalls,
+            // the loop like for (int i = 0; i < pendingCalls.size(); i++) can't stop.
```

Comment:
It seems to me the loop can stop but we don't want to verify the re-enqueue call right now, right?

Reply:
I think the loop can't stop, please see https://www.programiz.com/online-compiler/60n1G5xTAUTJO.

Yes, we don't want to verify re-enqueue call.

Reply:
oh, you are right. The re-added calls will encounter the same error again.